Start using the HtmlUtils safe template functions

andy-armstrong opened the pull request:
This change starts to convert the platform to use the new HtmlUtils class. I've primarily converted the Teams UI, along with the shared helper classes and Underscore templates that they depend upon.

## Sandbox
- [ ] https://andy-armstrong.sandbox.edx.org/

## Testing
- [ ] Unit, integration, acceptance tests as appropriate

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1: @cahrens 
- [ ] Code review #2: @robrap 

FYI: @AlasdairSwan @dan-f @dsjen @bjacobel 

## Post-review
- [ ] Squash commits

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

andy-armstrong commented:
This is a new version of https://github.com/edx/edx-platform/pull/11799 that is rebased against master.

andy-armstrong commented:
@robrap @cahrens I've addressed all of your feedback from the previous PR. Note that this is currently pointing at a GitHub hash until the UI Toolkit changes are approved and a release is cut.

robrap commented on common/static/common/js/components/views/feedback_notification.js:
```diff
@@ -1,34 +1,34 @@
+            // create Notification.Warning, Notification.Confirmation, etc
+            var capitalCamel, intents;
+            capitalCamel = _.compose(str.capitalize, str.camelize);
         intents = ["warning", "error", "confirmation", "announcement", "step-required", "help", "mini"];
```
Needs to match indent.

andy-armstrong replied:
Thanks. I'd tried to revert my changes in this file, but PyCharm fooled me as I had it in 'ignore whitespace' mode. I've fully reverted them now.

robrap commented on lms/djangoapps/teams/static/teams/js/spec/views/instructor_tools_spec.js:
```diff
@@ -29,29 +35,28 @@ define([
-                    'success'
+            expectSuccessMessage = function(view) {
+                var team = view.team;
+                expect($('.msg-content .copy').text().trim()).toContain(
```
Nit: you might not need trim() with toContain().

andy-armstrong replied:
You almost always need trim on text that you take out of the DOM. I'l try it without, but I wouldn't expect it to work.

robrap commented on lms/djangoapps/teams/static/teams/js/views/team_utils.js:
```diff
@@ -1,70 +1,70 @@
-                $('.teams-content .msg-content .copy').text(message);
-                messageElement.focus();
+                $message.removeClass('is-hidden').addClass(type);
+                HtmlUtils.setHtml($('.teams-content .msg-content .copy'), message);
```
@andy-armstrong I'll need help documenting and for future linting, knowing when $() can be called and when we need the new element() method.

andy-armstrong replied:
Note that this usage of ``$()`` isn't creating an element, but is finding a matching one. It is amazing all of the different behaviors that ``$`` supports based upon the type of arguments it receives.

The only usages that we need to switch out are when ``$()`` is used to create an element, e.g.

    $('<div></div>')

The idea is that ``HtmlUtils.element`` is safer because it only works with HTML snippets, i.e.

    HtmlUtils.element(HtmlUtils.HTML('<div></div>'));

robrap replied:
I still haven't seen HtmlUtils.element used in the wild, so I'm not sure how and where it will be safer.

robrap commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -44,7 +46,10 @@
-                    this.$el.html(_.template(teamsTemplate));
+                    HtmlUtils.setHtml(
+                        this.$el,
+                        HtmlUtils.template(teamsTemplate)()
```
Why the function call after template(...) (i.e. the second pair of `()`)?

andy-armstrong replied:
This is another piece of magic from JQuery. A template is a function, and if you call ``this.$el.html`` with a function then it will invoke it for you with no arguments, and then set the HTML to the result. This works smoothly with templates that don't need any options.

I considered doing the same thing with ``HtmlUtils.template`` but I don't like implementing different behaviors based upon the argument's type. It seemed clearer to me to actually invoke the template with no parameters. I did modify ``HtmlUtils.template`` so that if options aren't provided that it acts the same as if you'd passed an empty object.

robrap replied:
I thought I saw it used differently above, but don't see that now.  Thanks for the explanation.

robrap commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -32,9 +33,10 @@
                         TeamsView.prototype.render.call(self);
                         if (self.canUserCreateTeam()) {
-                            var message = interpolate_text(
+                            messageHtml = HtmlUtils.interpolateHtml(
```
As a best practice, would it make sense for this code to just move inside the template.  You then wouldn't need the call to ensureHtml() in the template, because this code would be right there.

andy-armstrong replied:
It seems like very complex logic to have live in a template. I prefer to minimize the logic in templates because they don't count towards code coverage so it is hard to know how well it is being tested. It is also much harder to debug code inside a template.

robrap replied:
Ok.  It's too bad i18n makes things so complicated.  It would be nice if all the html could be in the templates.

robrap commented on lms/djangoapps/teams/static/teams/templates/team-member.underscore:
```diff
@@ -1,7 +1,7 @@
 <li>
   <span class="team-member">
     <a class="member-profile" href="<%= memberProfileUrl %>">
```
You can do this later, but all `<%=` should be `<%-`, unless it contains an HtmlUtils function.

andy-armstrong replied:
Is it safe to HTML escape URLs? I wasn't sure what we should be doing yet.

robrap replied:
Yes.  In theory they should also be URL-escaped, and I'm not as clear on that, but HTML-escaping is right.

robrap commented on lms/djangoapps/teams/static/teams/templates/team-membership-details.underscore:
```diff
@@ -1,4 +1,4 @@
-<span class="member-count"><%= membership_message %></span>
+<span class="member-count"><%= HtmlUtils.ensureHtml(membership_message) %></span>
```
Another place where I wonder if it would make more sense to just have the message in here?

robrap commented on lms/static/js/components/card/views/card.js:
```diff
@@ -61,43 +62,47 @@
                         detail.setElement(detail.el).render();
                         detail.$el.addClass('meta-detail');
-                        detailsEl.append(detail.el);
+                        $details.append(detail.el);
```
Is this an append call that is becoming illegal?  What would it look like in the future?

andy-armstrong replied:
It isn't a problem because this is appending one DOM node to another. However, it may be a problem that our linter can't distinguish the two. Our new wrapper can't handle this case, at least as yet, so we can't convert it to anything else.

robrap commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -1,37 +1,42 @@
+            showNotificationMessage: function() {
+                var accountSettingsPageUrl = this.options.accountSettingsPageUrl,
+                    accountSettingsLinkHtml = HtmlUtils.joinHtml(
+                        HtmlUtils.HTML('<a href="' + accountSettingsPageUrl + '">'),
```
You should have an inner joinHtml() for this as well.

andy-armstrong replied:
That seems strange, because we are then implicitly HTML escaping a URL.

robrap commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -46,39 +51,39 @@
-                return interpolate_text(
-                    gettext("Profile image for {username}"), {username: this.model.get('username')}
+            imageAltText: function() {
+                return HtmlUtils.interpolateHtml(
```
StringUtils.

robrap commented on lms/static/js/student_profile/views/learner_profile_view.js:
```diff
@@ -1,71 +1,77 @@
+                    }
 
-                this.$('.profile-section-one-fields').append(this.options.usernameFieldView.render().el);
+                    this.$('.profile-section-one-fields').append(this.options.usernameFieldView.render().el);
```
Are all these appends ok?

andy-armstrong replied:
Yes, these are all element-to-element appends.

robrap replied:
Maybe the linter should allow append with anything that ends with ".el"?  Of course if this is put into a variable beforehand, that would break it.

robrap commented on lms/static/js/views/fields.js:
```diff
@@ -18,9 +20,17 @@
         var FieldViews = {};
+        var statusButtonHtml = function(label, classNames) {
+            return HtmlUtils.joinHtml(
+                HtmlUtils.HTML('<i class="fa ' + classNames + '" aria-hidden="true"></i>'),
```
in theory classNames should be escaped as well.

andy-armstrong replied:
These are static strings in the same file, so that seems unnecessary.

robrap replied:
The linter may look for this.  It's fine either way, but ultimately it may require a linter disable pragma.

robrap commented on lms/static/js/views/fields.js:
```diff
@@ -123,7 +136,8 @@
                 setTimeout(function () {
-                    if ((context === view.lastSuccessMessageContext) && (view.getNotificationMessage() === successMessage)) {
+                    if ((context === view.lastSuccessMessageContext) &&
+                        (view.getNotificationMessage() === successMessage.toString())) {
```
I guess this is one of those dangerous spots.  This should probably be mentioned in the docs.

robrap commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -25,10 +27,10 @@
+                 TeamMembersEditView, TeamProfileHeaderActionsView, TeamUtils, InstructorToolsView, teamsTemplate) {
             var TeamsHeaderModel = HeaderModel.extend({
                 initialize: function () {
                     _.extend(this.defaults, {nav_aria_label: gettext('Topics')});
```
Is this better here than directly in the template for some reason?

andy-armstrong replied:
I think the problem is that the header template is very generic, and so instance specific values like this need to be passed into it. The header template doesn't know that it is a header for topics, in this case.

robrap commented on lms/templates/components/header/header.underscore:
```diff
@@ -2,15 +2,15 @@
         <div class="sr-is-focusable" tabindex="-1"></div>
         <% if (breadcrumbs !== null && breadcrumbs.length > 0) { %>
-            <nav class="breadcrumbs" aria-label="<%- nav_aria_label %>">
+            <nav class="breadcrumbs" aria-label="<%= HtmlUtils.ensureHtml(nav_aria_label) %>">
```
Why did this change?  I don't think this is HTML, so why not use the `<%-`?  Also, if we moved the gettext() call here, it would be even for clear.

andy-armstrong replied:
Good catch. I've reverted it.

robrap commented on lms/templates/components/header/header.underscore:
```diff
@@ -2,15 +2,15 @@
+            <nav class="breadcrumbs" aria-label="<%= HtmlUtils.ensureHtml(nav_aria_label) %>">
             <% _.each(breadcrumbs, function (breadcrumb) { %>
-                <a class="nav-item" href="<%= breadcrumb.url %>"><%- breadcrumb.title %></a>
+                <a class="nav-item" href="<%= breadcrumb.url %>"><%= HtmlUtils.ensureHtml(breadcrumb.title) %></a>
```
`<%=` to `<%-` for at least breadcrumb.url, and possibly more.  Why ensureHtml() calls?

andy-armstrong replied:
There were situations where titles were allowed to pass HTML. It seemed more flexible to support both in a generic header template.

Again, I wasn't sure that HTML escaping URLs was a good thing to be doing.

robrap commented on lms/templates/fields/field_image.underscore:
```diff
@@ -1,17 +1,17 @@
 <div class="image-wrapper">
-    <img class="image-frame" src="<%- imageUrl %>"  alt="<%=imageAltText%>"/>
+    <img class="image-frame" src="<%- imageUrl %>"  alt="<%= HtmlUtils.ensureHtml(imageAltText) %>"/>
```
Again, would it be better to have the code for the text directly in the template and drop all of these ensureHtml() calls?

andy-armstrong replied:
This is another example of a very generic template being used in multiple instances.

robrap commented:
@andy-armstrong The code looks clean.  Thanks! 

I was tired when I reviewed, so I should double check.  This is a big review.  I dropped a bunch of comments that don't necessarily need to get refactored, but it would be nice to use them to help settle best practices to be documented.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/topic_card.js:
```diff
@@ -1,52 +1,62 @@
-                    )));
+                render: function() {
+                    var teamCount = this.model.get('team_count');
+                    HtmlUtils.setHtml(
```
@robrap Here's an example where we use ``HtmlUtils.setHtml`` with a string. It seems useful for updating simple paragraph and span tags. I couldn't find an example for ``HtmlUtils.append`` but it seems reasonable to support it for consistency.

robrap replied:
In our best practice guide, would we want people to use `this.$el.text(...)`?

andy-armstrong replied:
That's a very good point. I'll switch it...

robrap commented:
This is a general message being posted for all Safe Templates PRs.  If you are making steady progress, sorry about this, and thank you!

We need to get any Safe Template PRs merged as soon as possible.

Make sure you have the latest copy of `./scripts/safe_template_linter.py` and `./scripts/safe-commit-linter.sh`, and make sure that `./scripts/safe-commit-linter.sh` reports 0 violations for this PR.

Feel free to ping me with any questions, and/or refer to the [Safe Template Linter documentation](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#safe-template-linter ).

Thanks!

cahrens commented:
@andy-armstrong should this PR be closed?

andy-armstrong commented:
@cahrens No, I need to get back to merging this.

robrap commented:
@andy-armstrong FYI: When you rebase from master you'll get reports of violations for this PR in Jenkins under quality.

cahrens commented:
@andy-armstrong  OK to close this PR?

robrap commented:
@cahrens @andy-armstrong: I don't think so.  @scottrish will be reviewing the remaining safe template open PRs for next steps (prompted by the branch removal request).

What would it take to land this work?

andy-armstrong commented:
@robrap @cahrens I'll try rebasing it tonight and see how far off it seems.
